File: source/go/lowest_common_ancestor.go
package main

import "fmt"

type Node struct {
	value int
	left  *Node
	right *Node
}

func insert(root *Node, value int) *Node {
	if root == nil {
		return &Node{value: value}
	}
	if value < root.value {
		root.left = insert(root.left, value)
	} else if value > root.value {
		root.right = insert(root.right, value)
	}
	return root
}

func lca(root *Node, p, q int) *Node {
	if root == nil {
		return nil
	}
	if p < root.value && q < root.value {
		return lca(root.left, p, q)
	}
	if p > root.value && q > root.value {
		return lca(root.right, p, q)
	}
	return root
}

func main() {
	var root *Node
	values := []int{50, 30, 70, 20, 40, 60, 80}
	for _, v := range values {
		root = insert(root, v)
	}

	r1 := lca(root, 20, 40)
	fmt.Printf("lca(20, 40): %d\n", r1.value)

	r2 := lca(root, 20, 70)
	fmt.Printf("lca(20, 70): %d\n", r2.value)

	r3 := lca(root, 60, 80)
	fmt.Printf("lca(60, 80): %d\n", r3.value)
}

File: source/python/lowest_common_ancestor.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

def insert(root, value):
    if root is None:
        return Node(value)
    if value < root.value:
        root.left = insert(root.left, value)
    elif value > root.value:
        root.right = insert(root.right, value)
    return root

def lca(root, p, q):
    if root is None:
        return None
    if p < root.value and q < root.value:
        return lca(root.left, p, q)
    if p > root.value and q > root.value:
        return lca(root.right, p, q)
    return root

root = None
for v in [50, 30, 70, 20, 40, 60, 80]:
    root = insert(root, v)

r1 = lca(root, 20, 40)
print(f"lca(20, 40): {r1.value}")

r2 = lca(root, 20, 70)
print(f"lca(20, 70): {r2.value}")

r3 = lca(root, 60, 80)
print(f"lca(60, 80): {r3.value}")
